### PR TITLE
Use getaddrinfo(3) instead of gethostbyname(3)

### PR DESCRIPTION
**Pull request by alejandro-colomar**
gethostbyname(3) was removed in POSIX.1-2008.  It has been obsoleted,
and replaced by getaddrinfo(3), which is superior in several ways:

-  gethostbyname(3) is not reentrant.  There's a GNU extension,
   gethostbyname_r(3) which is reentrant, but it's not likely to be
   standardized for the following reason.  And we don't care too much
   about this point either.

-  gethostbyname(3) only supports IPv4, but getaddrinfo(3) supports both
   IPv4 and IPv6 (and may support other address families in the future).

We don't care about reentrancy, so for keeping the code simple (i.e.,
not touch call site to add code to free(3) an allocated buffer), I added
a static buffer for inet_ntop(3).  We could address that in the future,
but I don't think it's worth it.

BTW, we also replace inet_ntoa(3) by inet_ntop(3), as a consequence of
using getaddrinfo(3).  inet_ntoa(3) is also marked as deprecated, but
that deprecation seems to have been documented only in the manual page,
and POSIX doesn't mark it as deprecated.  The deprecation notice goes
back to when the inet_ntop(3) manual page was added by Sam Varshavchik
to the Linux man-pages in version 1.30 (year 2000).

So, this, apart from updating the code to POSIX.1-2008, is also adding
support for IPv6 :)  Although, probably many other parts of the code are
written for IPv4 only, so I wouldn't yet claim support for it.

A few notes:

-  I didn't check the return value of inet_ntop(3), since it can't fail
   for the given input:

   -  EAFNOSUPPORT:  We only call it with AF_INET and AF_INET6.
   -  ENOSPC:  We calculate the size of the buffer to be wide enough:
               MAX(INET_ADDRSTRLEN, INET6_ADDRSTRLEN) so it always fits.

Link: <https://github.com/shadow-maint/shadow/issues/600>
Cc: Dave Hagewood <admin@arrowweb.com>
Cc: Sam Varshavchik
Cc: Jakub Jelinek <jakub@redhat.com> @jakubjelinek
Cc: Iker Pedrosa <ipedrosa@redhat.com>
Signed-off-by: Alejandro Colomar <alx@kernel.org>

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
Rebased to master.

**Comment by alejandro-colomar** [comment]
Abort on IPv7 :)

**Comment by alejandro-colomar** [comment]
Changes:

- Rebased to master.
- Remove NITEMS() from this PR. (So, tests will fail, but the PR will be simpler, once the other is merged).

**Comment by alejandro-colomar** [comment]
Rebased to master.

File: src/login_nopam.c
Comment (hallyn) [inline src/login_nopam.c:277]: I assume this is a WIP, but abort()ing here is a change in behavior.

File: src/login_nopam.c
Comment (ikerexxe) [inline src/login_nopam.c]: I think it would make more sense to rename this variable. Something like `gai_error` would make more sense.

File: src/login_nopam.c
Comment (ikerexxe) [inline src/login_nopam.c]: Can you rename this variable?

File: libmisc/inet_sockaddr2str.c
Comment (ikerexxe) [inline libmisc/inet_sockaddr2str.c:27]: We are using SPDX to indicate the software license.

File: libmisc/inet_sockaddr2str.c
Comment (ikerexxe) [inline libmisc/inet_sockaddr2str.c:58]: You missed an `r` in `receiving`

**Review by ikerexxe — CHANGES_REQUESTED** [review]
There are some minor issues to address before getting my approval.

File: src/login_nopam.c
Comment (alejandro-colomar) [reply src/login_nopam.c:277]: No, it's not a WIP.

Yes, this is an important change in behavior, which is why I kept it separate from the cleanups for POSIX.1-2008:
<https://github.com/shadow-maint/shadow/pull/616#issuecomment-1362720236>.

The change in behavior (documented in the commit message) is:

The old gethostbyname(3) only supports IPv4.  The replacement, getaddrinfo(3) supports both IPv4 and IPv6 (and is not inherently limited to those, so a hypothetical IPv7 could also be supported in the future).  By changing the function, we're adding support for IPv6, at least in that part of the code.

Because the function is not inherently limited to IPv4/IPv6, I wanted to be cautious, and abort(3) on anything that's not 4 or 6.  So, if IPv7 is ever a thing, this code will gracefully abort (there's nothing better we can do), instead of doing harmful stuff.

Comment (alejandro-colomar) [reply src/login_nopam.c:277]: Although, I should have aborted in the first commit already.  I forgot to do that.  Will fix in a moment.

Comment (alejandro-colomar) [reply src/login_nopam.c:277]: Fixed.  Now I'm aborting since the first commit.

**Comment by alejandro-colomar** [comment]
Changes:

-  abort(3) in the first commit.

File: libmisc/inet_sockaddr2str.c
Comment (alejandro-colomar) [reply libmisc/inet_sockaddr2str.c:27]: Great.  I think I copied it from another file, so I'll check that.

File: libmisc/inet_sockaddr2str.c
Comment (alejandro-colomar) [reply libmisc/inet_sockaddr2str.c:58]: Thanks.

**Comment by alejandro-colomar** [comment]
Changes:

-  Use SPDX
-  Fix typo (missing r)

File: src/login_nopam.c
Comment (alejandro-colomar) [reply src/login_nopam.c]: Yep.  I used gai_err, to not make a line too long.

**Comment by alejandro-colomar** [comment]
Changes: Rename gai -> gai_err

**Comment by alejandro-colomar** [comment]
Changes: Rename p -> addr_str

File: src/login_nopam.c
Comment (alejandro-colomar) [reply src/login_nopam.c]: Yep

File: src/login_nopam.c
Comment (ikerexxe) [inline src/login_nopam.c:276]: You don't need to print this here as it was already printed in `inet_sockaddr2str()`.

Comment (alejandro-colomar) [reply src/login_nopam.c:276]: No, I didn't log in inet_sockaddr2str();  I only set errno.

File: src/login_nopam.c
Comment (hallyn) [reply src/login_nopam.c:277]: Hm, well, "gracefully abort" is an interesting phrase, but I see, if a site has user and no from table entries, the login should proceed, so it's ok.

Thanks.

File: src/login_nopam.c
Comment (alejandro-colomar) [reply src/login_nopam.c:276]: Ahh, I see where you got confused.  The log in `inet_sockaddr2str()` is for when no addresses are found.  This one is for when an IPv7? address is found.

**Comment by alejandro-colomar** [comment]
Changes: print strerror(); since I documented the meaning of errno for inet_sockaddr2str(), it's simpler to just print strerror().

File: src/login_nopam.c
Comment (hallyn) [inline src/login_nopam.c]: Just for readability, imo it would be nice if this used switch (sa->sa_family)

Comment (hallyn) [reply src/login_nopam.c]: Never mind - i see it doesn't matter :)

**Review by hallyn — APPROVED** [review]
Thanks.

**Comment by hallyn** [comment]
@ikerexxe  I'll wait for you to merge, as I don't know if you had some unfinished thoughts on this.  But I'm happy with it.

File: src/login_nopam.c
Comment (alejandro-colomar) [reply src/login_nopam.c]: Well, it can matter slightly for someone reading the diff, so I fixed it.  Anyway, it's a smaller diff now. :)

**Comment by alejandro-colomar** [comment]
Changes:
Changed an intermediate diff.  Nothing change in the global diff.

**Review by ikerexxe — APPROVED** [review]
Now that my minor comments were addressed I'm happy with the patches.

**Pull request merged**
